### PR TITLE
fix(gsd): stop false drift warnings and dead-ended discussions from scratch milestone state

### PR DESCRIPTION
**Pull request by jeremymcs**
## Intent

Fix two GSD runtime bugs diagnosed from a live test session (test-apps/890): (1) The startup drift check (checkMarkdownHierarchyAgainstDb) counted milestone dirs holding only CONTEXT/CONTEXT-DRAFT (no ROADMAP, no DB row) as markdown milestones. These are pre-registration discussion artifacts — the queued DB row is only inserted at discussion handoff — so every live or abandoned milestone discussion produced a false 'recovery-required' warning recommending '/gsd recover --confirm', an import that would materialize abandoned-discussion scratch dirs as ghost active milestones in the authoritative DB. Fix: filter such dirs from the comparison inside checkMarkdownHierarchyAgainstDb only, deliberately leaving the raw scans (countMarkdownHierarchy, recoverWouldDeleteDbRows — used by recover preflights, which must see what the importer would import) untouched. Deliberately did NOT change the md-importer or insert DB rows at ID-reservation time, to keep blast radius small. (2) A discussion interrupted by /clear after its CONTEXT artifact was saved became an immortal pending auto-start entry: the #3274 staleness heuristic in guided-flow requires the CONTEXT file to be ABSENT, so every subsequent /gsd dead-ended on 'Discussion already in progress — answer the question above' with no question above. Fix: deterministically clear the in-memory pending auto-start entry in the session_switch hook when reason is 'new' (/clear, /new destroy the conversation holding the interview); reason 'resume' keeps the entry because the restored transcript contains the question. The 30s heuristic is intentionally kept as a backstop for other failure modes. Auto-mode's own newSession() calls are safe because the handoff consumes the entry on agent_end before any dispatch. Tests added for both: 4 drift-check scenarios and a session_switch handler test (new clears, resume keeps). A third investigated symptom (duplicate 'M006 context written.' message) was confirmed to be the model restating its status across two turns — not a harness bug, no code change. NOTE: two previous runs of this head died on provider infrastructure errors, not findings: run 01KTQC2BD2Z683BEDBE60AM7H6 passed review with 0 findings and its test agent validated both fixes end-to-end (all 15 tests passed) before crashing on an API socket error; run 01KTQCY4EVJM28D2X468F53Y32 crashed mid-review stream. This is retry #3.

## What Changed

- Filter pre-registration discussion dirs (those holding only `CONTEXT`/`CONTEXT-DRAFT` with no `ROADMAP` and no queued DB row) out of the comparison inside `checkMarkdownHierarchyAgainstDb`, so abandoned or in-flight milestone discussions no longer trigger a false `recovery-required` warning recommending `/gsd recover --confirm`. The raw scans used by recover preflights are left untouched.
- Clear the in-memory pending auto-start entry in the `session_switch` hook when the reason is `new` (`/clear`, `/new`), preventing an interrupted post-CONTEXT discussion from becoming an immortal entry that dead-ends every subsequent `/gsd` on "Discussion already in progress"; `resume` keeps the entry since the restored transcript still holds the question.
- Add tests for both fixes (4 discussion-scratch drift scenarios, plus a session_switch handler test covering new-clears and resume-keeps) and document the `session_switch` clearing behavior in the new-milestone discuss-flow doc.

## Risk Assessment

✅ Low: Two tightly-scoped GSD bugfixes whose key safety invariants I verified against the surrounding code (newSession emits session_switch reason "new" but the handoff consumes the pending entry on agent_end first; basePath keying is consistent via ctx.cwd; raw drift scans left untouched), each backed by accurate targeted tests.

## Testing

The worktree arrived without node_modules or built package dist, so I installed deps and built packages plus test:compile as a setup prerequisite, then ran the two changed test files from dist-test with all 15 tests passing. Beyond the unit tests I wrote and ran a driver that invokes the actual production functions on a test-apps fixture and captured a CLI transcript: the startup check returns action none and reason in-sync with no recovery banner for discussion-scratch dirs while still flagging genuine drift, and the session_switch hook clears the pending handoff on reason new but keeps it on resume. No product failures were found; build-regenerated pi-ai generated files were reverted to keep the tree clean.

<details>
<summary>Evidence: End-to-end CLI transcript of both fixes against real production functions</summary>

```text
══════════════════════════════════════════════════════════════════════
 BUG 1 — startup drift check vs. live/abandoned discussion scratch dirs
══════════════════════════════════════════════════════════════════════
Scenario (test-apps/890): a healthy project (M001 in markdown AND DB)
plus three discussion artifacts mid-flight — dirs holding only CONTEXT
with NO ROADMAP and NO queued DB row (the row is inserted at handoff):
  • M002/M002-CONTEXT.md      (live discussion)
  • M003/M003-CONTEXT-DRAFT.md (abandoned draft)
  • M004/                      (empty scratch dir)

raw scanMarkdownHierarchy() sees milestones: [M001, M002, M003, M004]
  → roadmapless (scratch) dirs the recover preflight must still import: [M002, M003, M004]
  (raw scan deliberately UNCHANGED — recover preflights need to see these)

startup checkMarkdownHierarchyAgainstDb() → what the user's startup sees:
  action          : none
  reason          : in-sync
  recoveryCommand : (none — no banner shown)
  counted markdown: {"milestones":1,"slices":1,"tasks":1}

✅ RESULT: NO false 'recovery-required' warning, NO '/gsd recover --confirm'.
   Before the fix: action=recovery-required, banner urged /gsd recover,
   which would have imported M002/M003/M004 as ghost active milestones.

── sub-check: a REAL drift (markdown M001 tree, empty DB) still warns ──
  action=recovery-required  reason=db-empty  recoveryCommand=/gsd recover --confirm  markdown={"milestones":1,"slices":1,"tasks":1}
✅ Genuine drift is still caught — and the scratch dir did NOT inflate the count (M002 excluded).

══════════════════════════════════════════════════════════════════════
 BUG 2 — /clear must clear the immortal pending auto-start handoff entry
══════════════════════════════════════════════════════════════════════
Scenario: a discuss interview saved its CONTEXT artifact, then the user
ran /clear. The 30s staleness heuristic requires CONTEXT to be ABSENT,
so the in-memory handoff entry became immortal and every later /gsd
dead-ended on 'Discussion already in progress — answer the question
above' with no question above.

reason "new"   (/clear, /new): armed before=true  pending after=false
reason "resume"             : armed before=true  pending after=true

✅ RESULT: /clear (reason 'new') drops the stranded handoff entry, so the
   next /gsd starts clean instead of dead-ending. 'resume' keeps it because
   the restored transcript still contains the question.

All scenarios completed.
```
</details>
<details>
<summary>Evidence: Evidence driver source harness</summary>

```text
// Evidence driver — reproduces the live test-apps/890 scenario against the
// REAL production startup-drift functions and the REAL session_switch hook.
// Prints the banner/result an end user would actually experience.

import { mkdtempSync, mkdirSync, writeFileSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

import { ensureDbOpen } from "/Users/jeremymcspadden/.no-mistakes/worktrees/c2f4d848183f/01KTQD1641X8W5JFCVK613BFQ8/dist-test/src/resources/extensions/gsd/bootstrap/dynamic-tools.js";
import { closeDatabase, insertMilestone, insertSlice, insertTask } from "/Users/jeremymcspadden/.no-mistakes/worktrees/c2f4d848183f/01KTQD1641X8W5JFCVK613BFQ8/dist-test/src/resources/extensions/gsd/gsd-db.js";
import {
  checkMarkdownHierarchyAgainstDb,
  scanMarkdownHierarchy,
} from "/Users/jeremymcspadden/.no-mistakes/worktrees/c2f4d848183f/01KTQD1641X8W5JFCVK613BFQ8/dist-test/src/resources/extensions/gsd/migration-auto-check.js";
import { writeGSDDirectory } from "/Users/jeremymcspadden/.no-mistakes/worktrees/c2f4d848183f/01KTQD1641X8W5JFCVK613BFQ8/dist-test/src/resources/extensions/gsd/migrate/writer.js";
import { registerHooks } from "/Users/jeremymcspadden/.no-mistakes/worktrees/c2f4d848183f/01KTQD1641X8W5JFCVK613BFQ8/dist-test/src/resources/extensions/gsd/bootstrap/register-hooks.js";
import { setPendingAutoStart, clearPendingAutoStart, hasPendingAutoStart } from "/Users/jeremymcspadden/.no-mistakes/worktrees/c2f4d848183f/01KTQD1641X8W5JFCVK613BFQ8/dist-test/src/resources/extensions/gsd/pending-auto-start.js";

function scratch(base, id, file) {
  const dir = join(base, ".gsd", "milestones", id);
  mkdirSync(dir, { recursive: true });
  if (file) writeFileSync(join(dir, file), `# ${id} discussion context\n`);
}

const line = (s = "") => process.stdout.write(s + "\n");

// ───────────────────────────────────────────────────────────────────────────
line("══════════════════════════════════════════════════════════════════════");
line(" BUG 1 — startup drift check vs. live/abandoned discussion scratch dirs");
line("══════════════════════════════════════════════════════════════════════");
line("Scenario (test-apps/890): a healthy project (M001 in markdown AND DB)");
line("plus three discussion artifacts mid-flight — dirs holding only CONTEXT");
line("with NO ROADMAP and NO queued DB row (the row is inserted at handoff):");
line("  • M002/M002-CONTEXT.md      (live discussion)");
line("  • M003/M003-CONTEXT-DRAFT.md (abandoned draft)");
line("  • M004/                      (empty scratch dir)");
line("");

const base1 = mkdtempSync(join(tmpdir(), "evidence-drift-"));
try {
  await writeGSDDirectory(
    {
      projectContent: "# Legacy Project\n",
      decisionsContent: "",
      requirements: [],
      milestones: [
        {
          id: "M001", title: "Legacy Milestone", vision: "v", successCriteria: ["x"],
          research: null, boundaryMap: [],
          slices: [{
            id: "S01", title: "Legacy Slice", risk: "medium", depends: [], done: false,
            demo: "d", goal: "d", research: null, summary: null,
            tasks: [{ id: "T01", title: "Legacy Task", description: "", done: false, estimate: "", files: ["src/index.ts"], mustHaves: [], summary: null }],
          }],
        },
      ],
    },
    base1,
  );
  await ensureDbOpen(base1);
  insertMilestone({ id: "M001", title: "Legacy Milestone", status: "active" });
  insertSlice({ id: "S01", milestoneId: "M001", title: "Legacy Slice", status: "pending", risk: "medium", depends: [], demo: "d", sequence: 1 });
  insertTask({ id: "T01", sliceId: "S01", milestoneId: "M001", title: "Legacy Task", status: "pending" });

  scratch(base1, "M002", "M002-CONTEXT.md");
  scratch(base1, "M003", "M003-CONTEXT-DRAFT.md");
  scratch(base1, "M004");

  const raw = scanMarkdownHierarchy(base1);
  line(`raw scanMarkdownHierarchy() sees milestones: [${[...raw.milestones].sort().join(", ")}]`);
  line(`  → roadmapless (scratch) dirs the recover preflight must still import: [${[...raw.milestonesWithoutRoadmap].sort().join(", ")}]`);
  line("  (raw scan deliberately UNCHANGED — recover preflights need to see these)");
  line("");

  const result = await checkMarkdownHierarchyAgainstDb(base1);
  line("startup checkMarkdownHierarchyAgainstDb() → what the user's startup sees:");
  line(`  action          : ${result.action}`);
  line(`  reason          : ${result.reason}`);
  line(`  recoveryCommand : ${result.recoveryCommand ?? "(none — no banner shown)"}`);
  line(`  counted markdown: ${JSON.stringify(result.markdown)}`);
  line("");
  if (result.action === "none" && !result.recoveryCommand) {
    line("✅ RESULT: NO false 'recovery-required' warning, NO '/gsd recover --confirm'.");
    line("   Before the fix: action=recovery-required, banner urged /gsd recover,");
    line("   which would have imported M002/M003/M004 as ghost active milestones.");
  } else {
    line("❌ UNEXPECTED: a recovery banner was produced for scratch dirs.");
    process.exitCode = 1;
  }
} finally {
  closeDatabase();
  rmSync(base1, { recursive: true, force: true });
}

line("");
line("── sub-check: a REAL drift (markdown M001 tree, empty DB) still warns ──");
const base1b = mkdtempSync(join(tmpdir(), "evidence-drift-real-"));
try {
  await writeGSDDirectory(
    {
      projectContent: "# P\n", decisionsContent: "", requirements: [],
      milestones: [{
        id: "M001", title: "M001", vision: "v", successCriteria: ["x"], research: null, boundaryMap: [],
        slices: [{ id: "S01", title: "S", risk: "medium", depends: [], done: false, demo: "d", goal: "d", research: null, summary: null,
          tasks: [{ id: "T01", title: "T", description: "", done: false, estimate: "", files: ["a"], mustHaves: [], summary: null }] }],
      }],
    },
    base1b,
  );
  await ensureDbOpen(base1b); // DB left empty
  scratch(base1b, "M002", "M002-CONTEXT.md"); // scratch dir alongside the real drift
  const r = await checkMarkdownHierarchyAgainstDb(base1b);
  line(`  action=${r.action}  reason=${r.reason}  recoveryCommand=${r.recoveryCommand}  markdown=${JSON.stringify(r.markdown)}`);
  if (r.action === "recovery-required" && r.recoveryCommand === "/gsd recover --confirm") {
    line("✅ Genuine drift is still caught — and the scratch dir did NOT inflate the count (M002 excluded).");
  } else {
    line("❌ UNEXPECTED: real drift no longer detected.");
    process.exitCode = 1;
  }
} finally {
  closeDatabase();
  rmSync(base1b, { recursive: true, force: true });
}

// ───────────────────────────────────────────────────────────────────────────
line("");
line("══════════════════════════════════════════════════════════════════════");
line(" BUG 2 — /clear must clear the immortal pending auto-start handoff entry");
line("══════════════════════════════════════════════════════════════════════");
line("Scenario: a discuss interview saved its CONTEXT artifact, then the user");
line("ran /clear. The 30s staleness heuristic requires CONTEXT to be ABSENT,");
line("so the in-memory handoff entry became immortal and every later /gsd");
line("dead-ended on 'Discussion already in progress — answer the question");
line("above' with no question above.");
line("");

const base2 = mkdtempSync(join(tmpdir(), "evidence-autostart-"));
mkdirSync(join(base2, ".gsd", "milestones", "M001"), { recursive: true });
writeFileSync(join(base2, ".gsd", "milestones", "M001", "M001-CONTEXT.md"), "# M001 Context\n");

const handlers = new Map();
registerHooks({ on(ev, h) { handlers.set(ev, h); } }, []);
const sessionSwitch = handlers.get("session_switch");
const fakeCtx = { cwd: base2, ui: { notify: () => {}, setWidget: () => {}, setStatus: () => {} } };

async function fire(reason) {
  clearPendingAutoStart();
  setPendingAutoStart(base2, {
    basePath: base2, milestoneId: "M001",
    ctx: { ui: { notify: () => {} } }, pi: { sendMessage: () => {} },
  });
  const armed = hasPendingAutoStart(base2);
  try {
    await sessionSwitch({ type: "session_switch", reason, previousSessionFile: undefined }, fakeCtx);
  } catch { /* session plumbing may throw on the minimal fake ctx; clear runs first */ }
  return { armed, after: hasPendingAutoStart(base2) };
}

const newRun = await fire("new");
line(`reason "new"   (/clear, /new): armed before=${newRun.armed}  pending after=${newRun.after}`);
const resumeRun = await fire("resume");
line(`reason "resume"             : armed before=${resumeRun.armed}  pending after=${resumeRun.after}`);
line("");
if (newRun.armed && !newRun.after && resumeRun.armed && resumeRun.after) {
  line("✅ RESULT: /clear (reason 'new') drops the stranded handoff entry, so the");
  line("   next /gsd starts clean instead of dead-ending. 'resume' keeps it because");
  line("   the restored transcript still contains the question.");
} else {
  line("❌ UNEXPECTED: clear/keep behavior did not match the fix.");
  process.exitCode = 1;
}
clearPendingAutoStart();
rmSync(base2, { recursive: true, force: true });

line("");
line("All scenarios completed.");
```
</details>

## Pipeline

Updates from [git push no-mistakes](https://github.com/kunchenguid/no-mistakes)

<details>
<summary>✅ **intent** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Rebase** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Review** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Test** - passed</summary>

✅ No issues found.
- `migration-auto-check test 13 of 13 pass including 4 new discussion-scratch drift scenarios`
- `session-switch test 2 of 2 pass, new clears and resume keeps`
- `reproduction driver invoking real checkMarkdownHierarchyAgainstDb, scanMarkdownHierarchy, and the registered session_switch hook on a test-apps fixture, exit 0`
- `prerequisite build: pnpm install, build:pi, native build, test:compile`

</details>

<details>
<summary>✅ **Document** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Lint** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Push** - passed</summary>

✅ No issues found.

</details>

---

> [!NOTE]
> **Low Risk**
> Scoped startup-comparison and session-hook changes with tests; recover preflights and resume behavior are explicitly preserved.
> 
> **Overview**
> Fixes two GSD runtime edge cases around milestone discussion scratch state.
> 
> **Startup drift check:** `checkMarkdownHierarchyAgainstDb` now drops markdown milestone dirs that have no `ROADMAP` and no matching DB row from the comparison (pre-registration `CONTEXT` / `CONTEXT-DRAFT` only). That stops false `recovery-required` banners and `/gsd recover --confirm` during live or abandoned discussions. Raw `scanMarkdownHierarchy` / recover preflights are unchanged so importers still see scratch dirs.
> 
> **Pending discuss handoff:** On `session_switch` with `reason === "new"` (`/clear`, `/new`), `clearPendingAutoStart(basePath)` runs so an in-memory discuss→auto entry cannot outlive the cleared interview (especially after `CONTEXT` was written, when the 30s staleness heuristic no longer applies). `resume` leaves the entry intact.
> 
> Adds regression tests for both behaviors and documents the `/clear` behavior in the new-milestone discuss-flow doc.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7a306046b564499050aacdc0f1ccc1eb4d6ee1f7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
<a href="https://app.blacksmith.sh/open-gsd/codesmith/gsd-pi/pr/630"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1783642616&installation_id=134682257&pr_number=630&repository=open-gsd%2Fgsd-pi&return_to=https%3A%2F%2Fgithub.com%2Fopen-gsd%2Fgsd-pi%2Fpull%2F630&signature=c4df8ef9b95af4a6ee3d72648604c9ccbb37e8ca5d0c0428f7a4d97122efbb48"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: src/resources/extensions/gsd/bootstrap/register-hooks.ts [hunk 1]
```diff
@@ -817,6 +818,13 @@ export function registerHooks(
     clearDeferredApprovalGate();
     await resetAskUserQuestionsTurnCache();
     clearDiscussionFlowState(basePath);
+    // /clear or /new destroys the conversation holding a discuss interview, so
+    // its pending discuss→auto handoff can never be answered — clear it. Resume
+    // restores the interview transcript, so the entry survives. Auto-mode's own
+    // newSession() calls are safe: the handoff consumes the entry on agent_end.
+    if (event.reason === "new") {
+      clearPendingAutoStart(basePath);
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/bootstrap/register-hooks.ts:826]: **Accepted** — queued for a code change.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.

**Comment by jeremymcs** [comment]
Superseded: the code commits here landed via #629 (merged). The document-step addition was salvaged into a follow-up docs PR.

**Comment by jeremymcs** [comment]
**Code owner follow-up on Cursor Bugbot thread (comment #3384641649 — basePath key mismatch)**

**Bug confirmed and fixed.** Commit 61bd5f05 on `fix/discussion-scratch-drift`.

### What the bug was

`clearPendingAutoStart` in the `session_switch` handler used `contextBasePath(ctx)` (= raw `ctx.cwd`), but `setPendingAutoStart` entries are keyed by `resolveProjectRoot(ctx.cwd)` — which resolves worktree paths up to the main project root. When `ctx.cwd` is a milestone worktree path (e.g. `.gsd/worktrees/M001-branch`), the clear targeted the wrong map key and left the immortal entry untouched.

### Fix

```diff
- clearPendingAutoStart(basePath);
+ clearPendingAutoStart(resolveWorktreeProjectRoot(basePath));
```

`resolveWorktreeProjectRoot` is already imported at line 47. For non-worktree cwds it is an identity function, so the original test cases still pass unchanged.

### New test

Added *"reason 'new' from a milestone worktree cwd still drops the project-root entry"*: arms the entry under the project root, fires `session_switch` with `ctx.cwd` = the worktree subdir, asserts the entry is gone. All 3 tests pass.

### ⚠️ Follow-up PR needed

PR #629 (which superseded this PR) landed `clearPendingAutoStart(basePath)` **without** this worktree fix, so the dead-end is still present in `main`. The fix lives on `fix/discussion-scratch-drift` and requires a follow-up PR to land.
